### DISCUSSION TITLE
UX: Actions from context menus

### DISCUSSION THREAD

**Discussion by aragos** [bafolts/terraforming-mars#2121] [Ideas]
The current game UI requires a lot of screen space (i.e. either a large screen and/or much scrolling). There have been several suggestions around how to condense it (for example #2001). This is another such suggestion, focusing on *how to take actions*. It is an alternative to action UI suggestions such as #953.

## Click Elements to take Actions

Rather than having a separate area for actions, make actions flow from elements of the game that they relate to, via a click-invoked overlay menu. A few visual examples:

<img width="362" alt="Placing a tile on the map" src="https://user-images.githubusercontent.com/2575453/102144095-e4a15f80-3e32-11eb-89c4-3b13e819e77d.png">

<img width="415" alt="Placing a delegate on a party" src="https://user-images.githubusercontent.com/2575453/102143541-17972380-3e32-11eb-8989-f57756dd0f9c.png">

<img width="283" alt="Choosing between actions on a played card" src="https://user-images.githubusercontent.com/2575453/102143961-b91e7500-3e32-11eb-9367-a7dffe5c0b41.png">

<img width="683" alt="Selecting a card to play from hand, then choosing how to pay for it" src="https://user-images.githubusercontent.com/2575453/102144453-7f01a300-3e33-11eb-9037-cef7defa8d50.png">

## Further Benefits

Besides making the UI denser (which is my preference at least), triggering actions from existing UI elements makes it easier for the player to remain "in context". 

The current UI can require scrolling from the relevant information (say, the turmoil board) to the actions. The player must then keep in mind the necessary information (for example, distribution of delegates) to choose the desired action in the action menu (and the action menu must repeat information, such as the ruling bonus, that is already present on the turmoil board). 

### How to find Actions

One issue with this proposal is that "clickable" elements of the UI are not immediately apparent to the player. This can be improved by adding *visual highlights* to elements. In addition to highlighting where actions can be taken, the highlighting could also be used to _only_ indicate those actions which are available _at the moment_. For example if a player does not have sufficient heat resources to use them for increasing temperature - or if temperature is already at its max - do not highlight the UI elements that can trigger the "heat to temperature" action.

### Obscuring Game State

By using overlays this proposal risks obscuring game state when an overlay is active. One solution would be to move the context menu to a specially reserved area of the screen when active, similar to how the existing overview sidebar on the left works (for example at the bottom of the screen). I suspect however that "backing out" of the menu (even if more than one step is required, such as when choosing an action and *then* choosing how to pay for it) is generally fine - the decision tree is shallow enough that re-starting the action choosing flow after looking up some additional information does not cause a significant extra cost to the user.

If the action were to flow from a UI element where the necessary information is present, this memory overhead for the player can be avoided.

## Possible Action Mappings
A suggested mapping from actions to the things one can click on to trigger them. Note: Sub-lists here are supposed to show possible follow-on modal menus; "options" mean multiple choices in the top-level menu.

* Play project card
   * project card in hand 
      * choose payment distribution (if applicable) 
      * choose tile placement(s) (if applicable)
      * choose target card(s) (if applicable)
      * choose target player(s) (if applicable)
* Take action on a card ("red arrow")
   * card in "played" area (blue or corporation)
      * choose payment distribution (if applicable) 
      * choose tile placement(s) (if applicable)
      * choose target card(s) (if applicable)
      * choose target player(s) (if applicable)
* Trade with moon
   * moon tile (3 options, one for each payment type)
* Build colony
   * moon tile
* Send party delegate
   * target party (options: free or 5MC)
* Place greenery
   * target tile on map (options: standard project or plants)
   * own plant resources
      * choose target tile
* Increase temperature
   * temperature bar
   * own heat resources
* Place ocean
   * target tile on map
* Place city
   * target tile on map
* Increase energy production
   * own energy production/resources
* Sell patents
   * any project card in hand
* Increase Venus
   * venus track
* Claim milestone
   * target milestone
* Sponsor award
   * target award
* Pass
   * dedicated button
* Void second action
   * dedicated button
* Undo
   * dedicated button

I'm probably forgetting things but hopefully this gives a rough idea.

**Comment by kberg** [comment]
I love what you are thinking. This is fascinating. I can see a few areas of
concern, but it's how I might want to use such a tool.

>

**Comment by chosta** [comment]
First of all. Great suggestion. I reckon you have experience with these things to be bold enough to suggest such a feature. I have even more concerns than @kberg BUT if this is done right it might be the best thing that happened to mars since Opportunity. 
We do have to discuss some stuff and I propose a small PR with a really basic prototype so we can start somewhere.
My concerns:
* it's going to be a bug fiesta (I could be wrong, but probably not)
* mobile (many people use tablets and even phones to play)
* the visuals of the context menu could be improved (@ssimeonoff will agree)


> ### How to find Actions
> One issue with this proposal is that "clickable" elements of the UI are not immediately apparent to the player. This can be improved by adding _visual highlights_ to elements. In addition to highlighting where actions can be taken, the highlighting could also be used to _only_ indicate those actions which are available _at the moment_. For example if a player does not have sufficient heat resources to use them for increasing temperature - or if temperature is already at its max - do not highlight the UI elements that can trigger the "heat to temperature" action.

this is brilliant but it will take a lot of work, hope you are willing to put it :)

Sorry if I sound negative (I'm not). We just like to roast any new feature at the start so we make sure it's not a mistake.
You are welcome to discuss in discord too (I saw your comment there)

**Comment by sienmich** [comment]
I like this. I think this is the best solution for constant scrolling.
Currently there is this blinking highlight during tile selection, expanding it on all possible action would be nice.

**Comment by aragos** [comment]
Thanks for the feedback, I think it's all very reasonable. Luckily I have a bit of experience with web UI implementation even if it's outdated so hopefully this is something I can pull off. ;)

> * it's going to be a bug fiesta (I could be wrong, but probably not)

I expect that we'd want to test this in some way, yes - whether on a separate server or an option in-game. One thing that will aid us is that the basics of this kind of functionality are well supported by standard libraries. So the bugs will likely come from integration with the game engine. ;)

> * mobile (many people use tablets and even phones to play)

See above, I think this is where standard libraries will help a lot (for example with flow that fits on-screen etc). However I can also see utilizing the "bottom of the screen menu" as an alternative which would make it easier to do some of the layout especially on small form factors.
 
> * the visuals of the context menu could be improved (@ssimeonoff will agree)

The above examples are just mocks I whipped up very quickly in a non-graphics program. I wouldn't use that atrocious a color palette, borders, font, etc! I imagine that we can iterate on something once the necessary infrastructure is in place.

**Comment by kyshantry** [comment]
Super cool idea!! Strongly endorse (have no idea how to code, but as a player, would be great!)

**Comment by kevinb9n** [comment]
One of the things I have wanted most for the UI is to move more toward what I was calling "concreteness": doing things by interacting with the actual components in the places where they always live. And look here! You're thinking the same way, that's awesome!

Just to fill in some thoughts I was already having, but NOT to claim that they are better (they're probably not :-)):

THIS IS PROBABLY A TANGENT, if it is something we want to discuss then please LMK if it should be filed separately or what:

When I was thinking about the "action bar" UI (#953), one thing that occurred to me was that it may be possible for the same single UI to support both noun-first and verb-first interactions at the same time. Either click the action button first, then go click the component of choice, or start by selecting the component and everything in the action bar that doesn't apply to it goes gray. Well, and I would change the remaining button to be green or otherwise glow brightly to indicate that it's become the "do it" button at that point.

it's not *generally* a great idea to have two different ways to do the same thing though. It's just that each direction is good for different things. If I was already looking at the thing, like the colony, then it's nice to be able to click it first, especially because (crazy idea alert) the interface could allow me to select it BEFORE my turn so there is one fewer click to make once my timer runs. On the other hand, by clicking the action first I can maybe get it to highlight just the right set of game components that can work with it. (Also the thing I interact with first is a thing that represents my high-level breadth of options, and that's often useful)

I was making a drawing at the time so here's [a drawing about it](https://drive.google.com/file/d/1VvmLu_n0ntp9IfrXJgqOIXLoGYKs5Hzb/view?usp=sharing).
(Another component of the idea is that there should be a "status bar" in a well-defined location that always shows whatever the game is expecting to happen next. Another tangent not to discuss here, but just to understand what you're seeing in the image.)

My thoughts were no more detailed than that. And I don't have any yet on the context-menus idea because I *just* saw this.

Thanks so much for filing this!

**Comment by kevinb9n** [comment]
> the visuals of the context menu could be improved (@ssimeonoff will agree)

I suggest a convention when discussing interaction design or UI functionality of *intentionally* making things look cosmetically ugly. That's only *partly* so I can say I was doing that on purpose in my own mocks.

**Comment by kevinb9n** [comment]
> <img alt="Choosing between actions on a played card" width="283" src="https://user-images.githubusercontent.com/2575453/102143961-b91e7500-3e32-11eb-9367-a7dffe5c0b41.png">

Only for this part of it, we have this idea too: #2000 
I think it may be one notch further in the concreteness direction.
But, I had no answer for all the things like Power Infrastructure where some choice has to be made next.

**Reply by PedroRCSequeira** [reply]
the menu could have a trade 1 by default, with +- signs?

**Comment by kevinb9n** [comment]
Okay I felt weird having only three comments in a row here.

I figured out why I like this direction so much. The bug I just linked is about reducing scrolling – toward the hoped-for goal of a no-scrolling UI that players could find acceptable. And for that, space is at an extreme premium.

The job of the UI is both present information to the player and collect decisions from the player. Both important. But very differently shaped: the information part pretty much stays structurally constant through the whole game (tiles and delegates come and go, but the map and committee board – the structure – stay the same), whereas the interaction part is all ephemeral, short-lived, different for all kinds of different circumstances.

(Sure, your cards list grows and grows, but that's why I think it is the kind of thing that deserves its own separately scrollable panel, #2001 which you mentioned, which renders it structurally static after all.)

So, the more we are able to devote precious screen real estate to those durable elements (the information), the more stable everything feels; the little payment dialogs and things are situational anyway, so let them pop in and out.

(This may have all been perfectly clear to everyone else.)

To this part,

> One solution would be to move the context menu to a specially reserved area of the screen when active

I came to the same thought in a different direction. It might be nice if we don't need to do it, since space will be so very tight (to get to the no-scroll fever dream, at least). But if we take that approach, one thing I would like is for it to be a card-shaped area. :-) Because when the game needs to prompt me for which way I want to use my Viral Enhancers, I'd like it to do so by showing me the actual Viral Enhancers card.  (To be clear, I've been saying that concreteness means expressing decisions by interacting with actual game components in the places where they always live, but those two parts should be ranked, first part more important (the real component), second part less.)

**Comment by ssimeonoff** [comment]
Context menus for actions "Play card" and "Activate card" are a great idea, which we already discussed #983
When the Actions menu gets revised and the two lists "Cards in Hand" and "Play card" fuse into one (1 big list of cards means less UI and less scrolling), then it will be a great time to pick this up.

**Comment by bafolts** [comment]
I think this approach is the best way to start with heading towards a goal of not scrolling constantly. Should be easy to add this incrementally as well without impacting the entire UI in one step. Could eventually add this context menu more places. Starting with it on the board will most likely be simplest. As I plan to start focusing on the goal of UI scrolling next I may start with this context menu. Once the context menu were implemented for one area it should be easy to then add for other areas.

**Comment by PedroRCSequeira** [comment]
This looks awesome!

Regarding the "How to find Actions", could the old menu be kept separate? Or just keep a list of possible actions somewhere?

There were talks now about a "beginner/tutorial" mode, this could be on by default. For experienced players it shouldn't be an issue

**Comment by kberg** [comment]
This would be a terrific project.
